### PR TITLE
(#1566) Generify org.cactoos.list package

### PR DESCRIPTION
**Pull request by DmitryBarskov**
For #1566:

Generify some classes in org.cactoos.list package:
- org.cactoos.list.Immutable
- org.cactoos.list.ImmutableListIterator
- org.cactoos.list.Joined
- org.cactoos.list.JoinedListIterator

### CODE REVIEW

**Comment by DmitryBarskov** [comment]
Hey @victornoel! Could you have a look please?

**Comment by yegor256** [comment]
@victornoel what's up with this?

File: src/main/java/org/cactoos/list/JoinedListIterator.java [hunk 1]
```diff
@@ -47,7 +47,7 @@
     /**
      * {@link List} of {@link ListIterator}.
      */
-    private final List<ListIterator<T>> listiters;
+    private final List<ListIterator<? extends T>> listiters;
```
Comment (victornoel) [inline src/main/java/org/cactoos/list/JoinedListIterator.java:50]: @DmitryBarskov I think that should be `List<? extends ListIterator<...`, here and everywhere else :)

File: src/test/java/org/cactoos/list/JoinedListIteratorTest.java [hunk 2]
```diff
@@ -46,7 +46,7 @@ void joinsListIterators() {
         new Assertion<>(
             "Must concatenate iterable of listIterators together",
             new IterableOf<>(
-                new JoinedListIterator<>(
+                new JoinedListIterator<String>(
```
[the remark this answers is not in the document]
Comment (victornoel) [reply src/test/java/org/cactoos/list/JoinedListIteratorTest.java:49]: @DmitryBarskov I think let's keep the change for now, it's a bit more verbose but it works, while before, some other combinations would not work without the change :)

**Review by victornoel — CHANGES_REQUESTED** [review]
@DmitryBarskov late review :)

File: src/main/java/org/cactoos/list/JoinedListIterator.java [hunk 3]
```diff
@@ -84,15 +84,15 @@ public JoinedListIterator(final T item, final ListIterator<T> items) {
...
+    public JoinedListIterator(final ListIterator<? extends T> items, final T item) {
         this(new ListOf<>(items, new ListOf<>(item).listIterator()));
     }
 
     /**
      * Ctor.
      * @param items Items to concatenate
      */
-    public JoinedListIterator(final List<ListIterator<T>> items) {
+    public JoinedListIterator(final List<? extends ListIterator<? extends T>> items) {
```
Comment (DmitryBarskov) [inline src/main/java/org/cactoos/list/JoinedListIterator.java:95]:
```java
class Ints implements ListIterator<Integer> { /* ... */ }

new JoinedListIterator<>(
  Arrays.asList(new Ints(), new Ints())
);
```

**Comment by victornoel** [comment]
@DmitryBarskov thank you. Did you verify that there was no more class in the `list` package that would need changes like those?

**Comment by DmitryBarskov** [comment]
> @DmitryBarskov thank you. Did you verify that there was no more class in the `list` package that would need changes like those?

@victornoel yes, checked it twice

**Comment by victornoel** [comment]
@DmitryBarskov excellent, thank you very much!

**Comment by victornoel** [comment]
@rultor merge

**Pull request merged**